google/auth/aio/transport/sessions.py

```diff
@@ -0,0 +1,283 @@
+
+                Mind that it is not guaranteed that the timeout error is raised
+                at ``max_allowed_time``. It might take longer, for example, if
+                an underlying request takes a lot of time, but the request
+                itself does not timeout, e.g. if a large file is being
+                transmitted. The timout error will be raised after such
+                request completes.
+
+        Returns:
+                google.auth.aio.transport.Response: The HTTP response.
+
+        Raises:
+                google.auth.exceptions.TimeoutError: If the method does not complete within
+                the configured `max_allowed_time` or the request exceeds the configured
+                `timeout`.
+
+        # TODO (ohmayr): Investigate if this is required.
+        # I think it is reasonable to agree on a strict type for headers and
+        # let the caller handle any translation.
+            if headers:
+                for key in headers.keys():
+                    if type(headers[key]) is bytes:
+                        headers[key] = headers[key].decode("utf-8")
+
+        # Headers come in as bytes which isn't expected behavior, the resumable
+        # media libraries in some cases expect a str type for the header values,
+        # but sometimes the operations return these in bytes types.
+        """
+
+        retries = _exponential_backoff.AsyncExponentialBackoff(
+            total_attempts=transport.DEFAULT_MAX_REFRESH_ATTEMPTS
+        )
+        response = None
+        try:
+            async with timeout_guard(max_allowed_time) as with_timeout:
+                await with_timeout(
+                    # Note: before_request will attempt to refresh the credentials if expired.
+                    self._credentials.before_request(
+                        self._auth_request, method, url, headers
+                    )
+                )
+                async for _ in retries:
+                    response = await with_timeout(
+                        self._auth_request(
+                            url, method, data, headers, timeout, **kwargs
+                        )
+                    )
+                    if (
+                        response.status_code
+                        not in transport.DEFAULT_RETRYABLE_STATUS_CODES
+                    ):
+                        return response
+        except TimeoutError as exc:
+            raise exc
+        return response
+
+    async def get(
+        self,
+        url: str,
+        data: bytes = None,
+        headers: Mapping[str, str] = None,
+        max_allowed_time: Optional[float] = transport._DEFAULT_TIMEOUT_SECONDS,
+        timeout: float = transport._DEFAULT_TIMEOUT_SECONDS,
+        **kwargs,
+    ) -> transport.Response:
+        return await self.request(
+            "GET", url, data, headers, max_allowed_time, timeout, **kwargs
+        )
+
+    async def post(
```

Comment:
I think all of these functions will need a doc string.

Reply:
addressed.